planner/type_join.go

```diff
@@ -479,24 +424,47 @@ type typeJoinMany struct {
 	subSelect *mapper.Select
 }
 
-func (p *Planner) makeTypeJoinMany(
+func prepareScanNodeFilterForTypeJoin(
 	parent *selectNode,
 	source planNode,
 	subType *mapper.Select,
-) (*typeJoinMany, error) {
-	// split filter
-	if scan, ok := source.(*scanNode); ok {
-		var parentfilter *mapper.Filter
-		scan.filter, parentfilter = splitFilterByType(scan.filter, subType.Index)
-		if parentfilter != nil {
+) {
+	subType.ShowDeleted = parent.selectReq.ShowDeleted
+
+	scan, ok := source.(*scanNode)
+	if !ok || scan.filter == nil {
+		return
+	}
+
+	if filter.IsComplex(scan.filter) {
+		if parent.filter == nil {
+			parent.filter = new(mapper.Filter)
```

Comment:
nitpick: I do have a preference here for `parent.filter = mapper.NewFilter()` - the constructor makes it much easier to find places that instantiate new instances, and with the current code parent.filter can easily panic as `filter.Conditions` is nil.